test/StorageRent/StorageRent.t.sol

```diff
@@ -0,0 +1,1312 @@
+        uint256 price = fcStorage.price(1);
+        fcStorage.rent{value: price}(1, 1);
+
+        vm.prank(admin);
+        fcStorage.setVault(address(revertOnReceive));
+
+        vm.prank(treasurer);
+        vm.expectRevert(StorageRent.CallFailed.selector);
+        fcStorage.withdraw(price);
+    }
+
+    function testFuzzOnlyTreasurerCanWithdraw(address caller, uint256 amount) public {
+        vm.assume(caller != treasurer);
+
+        vm.prank(caller);
+        vm.expectRevert(StorageRent.NotTreasurer.selector);
+        fcStorage.withdraw(amount);
+    }
+
+    function continuousCredit(uint256 start, uint256 end, uint256 units) public {
+        uint256 rented = fcStorage.rentedUnits();
+        uint256 len = end - start;
+        uint256 totalUnits = len * units;
+        vm.assume(totalUnits <= fcStorage.maxUnits() - fcStorage.rentedUnits());
+
+        // Expect emitted events
+        for (uint256 i; i < len; ++i) {
+            vm.expectEmit(true, true, false, true);
+            emit Rent(operator, start + i, units);
+        }
+        vm.prank(operator);
+        fcStorage.continuousCredit(start, end, units);
+
+        // Expect rented units to increase
+        assertEq(fcStorage.rentedUnits(), rented + totalUnits);
+    }
+
+    function batchCredit(uint256[] memory ids, uint256 units) public {
+        uint256 rented = fcStorage.rentedUnits();
+        uint256 totalUnits = ids.length * units;
+        vm.assume(totalUnits <= fcStorage.maxUnits() - fcStorage.rentedUnits());
+
+        // Expect emitted events
+        for (uint256 i; i < ids.length; ++i) {
+            vm.expectEmit(true, true, false, true);
+            emit Rent(operator, ids[i], units);
+        }
+        vm.prank(operator);
+        fcStorage.batchCredit(ids, units);
+
+        // Expect rented units to increase
+        assertEq(fcStorage.rentedUnits(), rented + totalUnits);
+    }
+
+    function batchRentStorage(
+        address msgSender,
+        uint256[] memory ids,
+        uint256[] memory units
+    ) public returns (uint256) {
+        uint256 rented = fcStorage.rentedUnits();
+        uint256 totalUnits;
+        for (uint256 i; i < units.length; ++i) {
+            totalUnits += units[i];
+        }
+        uint256 totalCost = fcStorage.price(totalUnits);
+        vm.deal(msgSender, totalCost);
+        vm.assume(totalUnits <= fcStorage.maxUnits() - fcStorage.rentedUnits());
+
+        // Expect emitted events
+        for (uint256 i; i < ids.length; ++i) {
+            if (units[i] != 0) {
+                vm.expectEmit(true, true, false, true);
+                emit Rent(msgSender, ids[i], units[i]);
+            }
+        }
+        vm.prank(msgSender);
+        fcStorage.batchRent{value: totalCost}(ids, units);
+
+        // Expect rented units to increase
+        assertEq(fcStorage.rentedUnits(), rented + totalUnits);
+        return totalCost;
+    }
+
+    function credit(address msgSender, uint256 id, uint256 units) public {
```

Comment:
for the test methods `credit` and `rentStorage` I was a little confused about what they did when I read them in code. 

their purpose is to assert that a certain function works with side effects, so maybe we should call them `assertCredit` and `assertRentStorage` so that their purpose is more clear?